### PR TITLE
Bigipdriver.py hangs when exceptions occur

### PR DESCRIPTION
**Pull request by russokj**
Problem: If the CCCL library throws a Big-IP exception (or any
exception), or if the driver encounters an exception while reading
the config file, then the driver hangs due to the thread terminating
(but not the process).

Solution: Catch the exceptions, log an error as specific as possible,
and then enter the retry loop.

Fixes: #506

### CODE REVIEW

File: python/bigipconfigdriver.py [hunk 1]
```diff
@@ -317,50 +314,24 @@ def _do_reset(self):
...
-                        log.error("CCCL Error: %s", e.msg)
-                        raise e
+                try:
+                    config = _parse_config(self._config_file)
+                    # No 'resources' indicates that the controller is not
+                    # yet ready -- it does not mean to apply an empty config
+                    if 'resources' not in config:
+                        continue
+                    incomplete = self._update_cccl(config)
+                except ValueError:
```
Comment (edarzins) [inline python/bigipconfigdriver.py]: Is this the only exception that can occur here? CCCL is supposed to catch all BIG-IP/SDK related exceptions, but if it misses something, we'll still have the same problem.

File: python/bigipconfigdriver.py [hunk 2]
```diff
@@ -401,6 +372,53 @@ def _do_reset(self):
...
+
+                # Manually delete custom profiles (if needed)
+                if customProfiles and \
+                        mgr.get_schema_type() == 'ltm':
+                    _delete_unused_ssl_profiles(
+                        mgr.mgmt_root(),
+                        partition,
+                        cfg_ltm)
+
+            except F5CcclError as e:
```
Comment (f5yacobucci) [inline python/bigipconfigdriver.py]: The stacktrace I got from Mark's runs showed the last frame raising ConnectionError. Are we sure that will be converted to an F5CcclError? We may need to catch ConnectionError as well.

His stacktrace is in the comment of issue 506.

File: python/bigipconfigdriver.py [hunk 3]
```diff
@@ -317,50 +314,24 @@ def _do_reset(self):
...
+                        continue
+                    incomplete = self._update_cccl(config)
+                except ValueError:
+                    formatted_lines = traceback.format_exc().splitlines()
+                    last_line = formatted_lines[-1]
+                    log.error('Failed to process the config file {} ({})'
+                              .format(self._config_file, last_line))
+                    incomplete = 1
+                # TODO(kenr): Must update unit tests when catching this exception
+                # except Exception:
```
Comment (russokj) [inline python/bigipconfigdriver.py]: This chunk of code will be enabled once I get the unit tests working, so it will catch all other exceptions.

**Comment by russokj** [comment]
Before this change, an exception would have the following signature in the controller logs:
```
21:55:44 2018/01/23 04:53:31 [INFO] Traceback (most recent call last):
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/threading.py", line 804, in __bootstrap_inner
21:55:44 2018/01/23 04:53:31 [INFO]     self.run()
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/threading.py", line 757, in run
21:55:44 2018/01/23 04:53:31 [INFO]     self.__target(*self.__args, **self.__kwargs)
21:55:44 2018/01/23 04:53:31 [INFO]   File "/app/python/bigipconfigdriver.py", line 320, in _do_reset
21:55:44 2018/01/23 04:53:31 [INFO]     config = _parse_config(self._config_file)
21:55:44 2018/01/23 04:53:31 [INFO]   File "/app/python/bigipconfigdriver.py", line 584, in _parse_config
21:55:44 2018/01/23 04:53:31 [INFO]     config_json = json.load(config)
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/json/__init__.py", line 291, in load
21:55:44 2018/01/23 04:53:31 [INFO]     **kw)
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/json/__init__.py", line 339, in loads
21:55:44 2018/01/23 04:53:31 [INFO]     return _default_decoder.decode(s)
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/json/decoder.py", line 364, in decode
21:55:44 2018/01/23 04:53:31 [INFO]     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
21:55:44 2018/01/23 04:53:31 [INFO]   File "/opt/rh/python27/root/usr/lib64/python2.7/json/decoder.py", line 382, in raw_decode
21:55:44 2018/01/23 04:53:31 [INFO]     raise ValueError("No JSON object could be decoded")
21:55:44 2018/01/23 04:53:31 [INFO] ValueError: No JSON object could be decoded
21:55:44 2018/01/23 04:53:31 [INFO] [2018-01-23 04:53:31,065 __main__ INFO] entering inotify loop to watch /tmp/k8s-bigip-ctlr.config553021570/config.json
21:55:44 2018/01/23 04:53:31 [INFO] 
```

Now, a json error will result in the following (big-ip errors would have a similar output):
```
2018/01/24 13:24:44 [ERROR] [2018-01-24 13:24:44,749 __main__ ERROR] Failed to process the config file /tmp/k8s-bigip-ctlr.config054120764/config.json (ValueError: No JSON object could be decoded)
2018/01/24 13:24:44 [ERROR] [2018-01-24 13:24:44,749 __main__ ERROR] Error applying config, will try again in 1 seconds
2018/01/24 13:24:45 [WARNING] Overwriting existing entry for backend {ServiceName:pool-iapp-svc ServicePort:80 Namespace:default}
2018/01/24 13:24:45 [INFO] Wrote 0 Virtual Server and 1 IApp configs
...
```

Unknown errors will have a little more information:
```
2018/01/24 13:25:11 [ERROR] [2018-01-24 13:25:11,465 __main__ ERROR] Unexpected error
2018/01/24 13:25:11 [INFO] Traceback (most recent call last):
2018/01/24 13:25:11 [INFO]   File "/app/python/bigipconfigdriver.py", line 322, in _do_reset
2018/01/24 13:25:11 [INFO]     if 'resources' not in config:
2018/01/24 13:25:11 [INFO] TypeError: argument of type 'NoneType' is not iterable
2018/01/24 13:25:11 [ERROR] [2018-01-24 13:25:11,465 __main__ ERROR] Error applying config, will try again in 1 seconds
2018/01/24 13:25:12 [INFO] Wrote 0 Virtual Server and 1 IApp configs
...
```
The key takeway is that the python driver continues to process config changes.

**Pull request merged**